### PR TITLE
Allow per-source status line customization

### PR DESCRIPTION
**Pull request by JCavallo**
Proposition for https://github.com/Shougo/denite.nvim/issues/407

### CODE REVIEW

**Comment by Shougo** [comment]
I think `doc/denite.txt` should be updated.

File: rplugin/python3/denite/source/base.py [hunk 1]
```diff
@@ -49,3 +49,6 @@ def gather_candidate(self, context):
 
     def debug(self, expr):
         denite.util.debug(self.vim, expr)
+
+    def get_status_line(self):
```
Comment (Shougo) [inline rplugin/python3/denite/source/base.py]: I think `get_statusline()` or `get_status()` is better.

**Comment by JCavallo** [comment]
I was not sure if the doc was needed, it is ok now

File: rplugin/python3/denite/source/base.py [hunk 2]
```diff
@@ -49,3 +49,6 @@ def gather_candidate(self, context):
 
     def debug(self, expr):
         denite.util.debug(self.vim, expr)
+
+    def get_statusline(self):
```
Comment (Shougo) [inline rplugin/python3/denite/source/base.py]: Please change it to `def get_statusline(self, context)`.
It should be.

File: rplugin/python3/denite/ui/default.py [hunk 3]
```diff
@@ -306,10 +306,11 @@ def update_candidates(self):
...
         sources = ''
         self._candidates = []
-        for name, entire, partial, patterns in self._denite.filter_candidates(
-                self._context):
+        for source, entire, partial, patterns in (
+                self._denite.filter_candidates(self._context)):
             self._candidates += partial
-            sources += '{}({}/{}) '.format(name, len(partial), len(entire))
+            sources += '{}({}/{}) '.format(
+                source.get_statusline(), len(partial), len(entire))
```
Comment (Shougo) [inline rplugin/python3/denite/ui/default.py]: Please change it to `source.get_statusline(self._context)`.

File: doc/denite.txt [hunk 4]
```diff
@@ -1867,6 +1867,12 @@ on_close	(Function)			(Optional)
 		It is called before quit.
 		It takes {self} and {context} as its parameter.
 
+					*denite-source-attribute-get_statusline*
+get_statusline	(Function)			(Optional)
+		It is called to calculate the status line for the source
```
Comment (Shougo) [inline doc/denite.txt]: `It takes {self} and {context} as its parameter.`

**Comment by Shougo** [comment]
I think `get_statusline()` should take `context` argument.

**Comment by JCavallo** [comment]
Good idea. Done

**Comment by Shougo** [comment]
Please wait to merge.
I have not much time now.

**Comment by JCavallo** [comment]
No problem, take your time

**Comment by Shougo** [comment]
Hm.  I should change the name `get_statusline()` to `get_status()` later.
Source can return `status`, it is displayed in statusline.
But it is just implementation.

**Comment by Shougo** [comment]
Merged.
And I have improved the implementation.

**Comment by JCavallo** [comment]
Thanks a lot

**Pull request merged**